File: 2001-number-of-pairs-of-interchangeable-rectangles/2001-number-of-pairs-of-interchangeable-rectangles.java
class Solution {
    public long interchangeableRectangles(int[][] rectangles) {
        long count = 0;
        int last = rectangles.length-1;
        HashMap<Double, Integer> map = new HashMap<>();
        map.put(Double.valueOf(rectangles[last][0]) / Double.valueOf(rectangles[last][1]),1);
        
        for(int i = last-1; i >=0; --i){
            Double current = Double.valueOf(rectangles[i][0]) / Double.valueOf(rectangles[i][1]);
            count += map.getOrDefault(current, 0);
            map.put(current, map.getOrDefault(current, 0) + 1);
        }
        
        return count;
    }
}

File: 2001-number-of-pairs-of-interchangeable-rectangles/2001-number-of-pairs-of-interchangeable-rectangles.py
from collections import defaultdict
class Solution:
    def interchangeableRectangles(self, rectangles: List[List[int]]) -> int:
        # Use a map to keep track of all the rectangles with the same ratio
        # Use the combination formula to get the number of combinations for a given ratio
        # (n!) / ((n - k!) * k!) 
        # where n is the count for a given ratio and k is the size of an individual combo (2 in this case)
        ratio_map = defaultdict(int)
        num_pairs = 0
        for w, h in rectangles:
            ratio_map[w/h] += 1
            
            
        
        for val in ratio_map.values():
            if val > 1:
                num_pairs += math.factorial(val) // (math.factorial(val - 2) * math.factorial(2))
        
        return num_pairs